Factor backend routing logic out of convolution forward

jbschlosser opened the pull request:
This PR introduces a new function `_select_conv_backend` that returns a `ConvBackend` enum representing the selected backend for a given set of convolution inputs and params.

The function and enum are exposed to python for testing purposes through `torch/csrc/Module.cpp` (please let me know if there's a better place to do this).

A new set of tests validates that the correct backend is selected for several sets of inputs + params. Some backends aren't tested yet:
* nnpack (for mobile)
* xnnpack (for mobile)
* winograd 3x3 (for mobile)

Some flowcharts for reference:
![conv_routing_graph md](https://user-images.githubusercontent.com/75754324/140828957-1135b400-38c0-4c9f-87ef-4f33ceebeeae.png)
![conv_nogroup_routing_graph md](https://user-images.githubusercontent.com/75754324/140828977-ed223a4e-aa86-49f1-9925-c0f6b9ab36af.png)

jbschlosser commented on aten/src/ATen/native/ConvUtils.h:
```diff
@@ -5,6 +5,76 @@
 
 namespace at { namespace native {
 
+// NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init)
+struct ConvParams {
+  std::vector<int64_t> stride;
+  std::vector<int64_t> padding;
+  std::vector<int64_t> dilation;
+  bool transposed;
+  std::vector<int64_t> output_padding;
+  int groups;
+  bool benchmark;
+  bool deterministic;
+  bool cudnn_enabled;
+  bool allow_tf32;
+
+  bool is_strided() const;
+  bool is_dilated() const;
+  bool is_padded() const;
+  bool is_output_padding_neg() const;
+  bool is_output_padding_big() const;
+  bool is_padding_neg() const;
+  bool is_stride_nonpos() const;
+  void view1d_as_2d();
+  bool use_cpu_depthwise3x3_winograd(const at::Tensor& input, const at::Tensor& weight, const at::Tensor& bias) const;
+  bool needs_64bit_indexing_no_split(const at::Tensor& input, const at::Tensor& weight) const;
+  bool use_cudnn(const at::Tensor& input, const at::Tensor& weight) const;
+  bool use_cudnn_depthwise(const at::Tensor& input, const at::Tensor& weight) const;
+  bool use_miopen(const at::Tensor& input, const at::Tensor& weight, bool bias_defined) const;
+  bool use_mkldnn(const at::Tensor& input, const at::Tensor& weight) const;
+  bool use_nnpack(const at::Tensor& input, const at::Tensor& weight) const;
+  bool use_xnnpack(const at::Tensor& input, const at::Tensor& weight, const at::Tensor& bias) const;
+  bool is_depthwise(const at::Tensor& input, const at::Tensor& weight) const;
+};
```
Moved here from `Convolution.cpp`

jbschlosser commented on aten/src/ATen/native/ConvUtils.h:
```diff
@@ -5,6 +5,76 @@
+  Xnnpack2d
+};
+
+// Function to select the convolution backend based on the inputs and params.
+// This overload is used internally.
+TORCH_API ConvBackend _select_conv_backend(
+    const Tensor& input,
+    const Tensor& weight,
+    const Tensor& bias,
+    const ConvParams& params);
+
+// Overload for selecting the convolution backend from the full set of convolution inputs.
+// This overload is exposed to python for testing, etc.
+TORCH_API ConvBackend _select_conv_backend(
+    const Tensor& input, const Tensor& weight, const c10::optional<Tensor>& bias_opt,
+    IntArrayRef stride, IntArrayRef padding, IntArrayRef dilation,
+    bool transposed, IntArrayRef output_padding, int64_t groups);
```
There's two overloads right now, which isn't great. The one exposed to python purposefully matches the signature of `convolution`

jbschlosser commented on torch/csrc/Module.cpp:
```diff
@@ -948,6 +949,36 @@ Call this whenever a new thread is created in order to propagate values from
     }))
     .def("expired", &WeakTensorRef::expired);
 
+  py::enum_<at::native::ConvBackend>(py_module, "ConvBackend")
+    .value("CudaDepthwise2d", at::native::ConvBackend::CudaDepthwise2d)
+    .value("CudaDepthwise3d", at::native::ConvBackend::CudaDepthwise3d)
+    .value("Cudnn", at::native::ConvBackend::Cudnn)
+    .value("CudnnTranspose", at::native::ConvBackend::CudnnTranspose)
+    .value("Empty", at::native::ConvBackend::Empty)
+    .value("Miopen", at::native::ConvBackend::Miopen)
+    .value("MiopenDepthwise", at::native::ConvBackend::MiopenDepthwise)
+    .value("MiopenTranspose", at::native::ConvBackend::MiopenTranspose)
+    .value("Mkldnn", at::native::ConvBackend::Mkldnn)
+    .value("MkldnnEmpty", at::native::ConvBackend::MkldnnEmpty)
+    .value("NnpackSpatial", at::native::ConvBackend::NnpackSpatial)
+    .value("Overrideable", at::native::ConvBackend::Overrideable)
+    .value("Slow2d", at::native::ConvBackend::Slow2d)
+    .value("Slow3d", at::native::ConvBackend::Slow3d)
+    .value("SlowDilated2d", at::native::ConvBackend::SlowDilated2d)
+    .value("SlowDilated3d", at::native::ConvBackend::SlowDilated3d)
+    .value("SlowTranspose2d", at::native::ConvBackend::SlowTranspose2d)
+    .value("SlowTranspose3d", at::native::ConvBackend::SlowTranspose3d)
+    .value("Winograd3x3Depthwise", at::native::ConvBackend::Winograd3x3Depthwise)
+    .value("Xnnpack2d", at::native::ConvBackend::Xnnpack2d);
+
+  py_module.def("_select_conv_backend", [](
+        const at::Tensor& input, const at::Tensor& weight, const c10::optional<at::Tensor>& bias_opt,
+        at::IntArrayRef stride_, at::IntArrayRef padding_, at::IntArrayRef dilation_,
+        bool transposed_, at::IntArrayRef output_padding_, int64_t groups_) {
+      return at::native::_select_conv_backend(
+          input, weight, bias_opt, stride_, padding_, dilation_, transposed_, output_padding_, groups_);
+  });
```
Exposes `ConvBackend` and `_select_conv_backend` to python

albanD commented on aten/src/ATen/native/ConvUtils.h:
```diff
@@ -5,6 +5,76 @@
+
+// Function to select the convolution backend based on the inputs and params.
+// This overload is used internally.
+TORCH_API ConvBackend _select_conv_backend(
```
nit nothing else in this file is pre-pended with `_` so this one doesn't really need to either I guess?

jbschlosser replied:
If I don't prepend with underscore here, it fails the test in `test_public_bindings.py` due to exposing it to python for testing

albanD commented on test/test_nn.py:
```diff
@@ -13261,6 +13263,134 @@ def test_conv3d_valid_padding_backward(self, device):
+        # === mkldnn ===
+        # 1D conv is not supported for mkldnn tensors
+        subtest(((2, 6, 7), False, False, 3, torch._mkldnn, torch._C.ConvBackend.Mkldnn),
+                decorators=[onlyCPU, skipCPUIfNoMkldnn, expectedFailure('cpu')], name='mkldnn1d'),
```
expected failure?

jbschlosser replied:
Found that 1D mkldnn inputs fail with the message: "opaque tensors do not have strides".

An explicit check for this case should be added - it'd be nice to have `assertRaisesRegex` in decorator form for testing this

albanD replied:
Ok!
Maybe link the corresponding issue here in case a future reader is wondering.

jbschlosser replied:
Linked the issue! #68034

albanD commented on test/test_nn.py:
```diff
@@ -13261,6 +13263,134 @@ def test_conv3d_valid_padding_backward(self, device):
+        subtest(((0, 0, 7, 8, 9), False, False, 3, torch._mkldnn, torch._C.ConvBackend.MkldnnEmpty),
+                decorators=[onlyCPU, skipCPUIfNoMkldnn], name='mkldnn_empty_batch_channel3d'),
+    ])
+    @parametrize_test("strided", [False, True])
```
What is this parametrize for?

jbschlosser replied:
This one tests both the stride=1 and stride > 1 cases for all of the above

albanD commented on test/test_nn.py:
```diff
@@ -13261,6 +13263,134 @@ def test_conv3d_valid_padding_backward(self, device):
+    def test_conv_backend_selection(
+            self, device, input_shape, strided, transposed, dilated, groups, layout, backend_expected):
+        # Build up inputs.
+        dtype = torch.float32
```
No dtype specific impls? :o Good news (I guess)

jbschlosser replied:
Some backends are not available for certain dtypes, but I've left the more complicated routing logic to future tests.. there is a huge combinatorial explosion of various scenarios that could be tested.

albanD commented on test/test_nn.py:
```diff
@@ -13261,6 +13263,134 @@ def test_conv3d_valid_padding_backward(self, device):
         self.assertEqual(gy_expect, gy_actual)
 
+    @skipMeta
+    @parametrize_test("input_shape,transposed,dilated,groups,layout,backend_expected", [
```
Can we have a comment here which backends are not tested?

albanD commented on test/test_public_bindings.py:
```diff
@@ -54,6 +54,7 @@ def test_no_new_bindings(self):
             "ConcreteModuleType",
             "ConcreteModuleTypeBuilder",
             "CONV_BN_FUSION",
+            "ConvBackend",
```
You most likely don't want this right?

jbschlosser replied:
should I prepend an underscore then to pass the test? `_ConvBackend`

albanD replied:
Yes I think so. Unless you want to make this public API and document it? :D

albanD reviewed (commented):
Sounds good. Only some comments and spurious public API.

zou3519 commented:
> Some backends aren't tested yet:
> 
> miopen (for AMD ROCm)
> nnpack (for mobile)
> xnnpack (for mobile)
> winograd 3x3 (for mobile)

I just started reading through this, but what is the plan for those? Is it fine to ignore them in this PR or do they need to be handled here?

zou3519 commented on aten/src/ATen/native/ConvUtils.h:
```diff
@@ -5,6 +5,76 @@
+  std::vector<int64_t> dilation;
+  bool transposed;
+  std::vector<int64_t> output_padding;
+  int groups;
```
nit: You don't need to fix this because this is a move, but this looks like a bug: groups seems like it should be int64_t?

jbschlosser replied:
> > Some backends aren't tested yet:
> > miopen (for AMD ROCm)
> > nnpack (for mobile)
> > xnnpack (for mobile)
> > winograd 3x3 (for mobile)
> 
> I just started reading through this, but what is the plan for those? Is it fine to ignore them in this PR or do they need to be handled here?

Current plan is to ignore them for this PR but add a note that they should be handled at some point. I was able to get miopen tested, so that just leaves the mobile backends. I can open a mobile issue indicating that they're not tested and leave it to those folks to deal with it.

jbschlosser replied:
Yeah I agree with you - seems like a bug. Think it's worth opening an issue for?

zou3519 commented on aten/src/ATen/native/ConvUtils.h:
```diff
@@ -5,6 +5,76 @@
+};
+
+// Function to select the convolution backend based on the inputs and params.
+// This overload is used internally.
```
nit: "internally" is a overloaded word (might mean internally at fb)

zou3519 commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -817,6 +785,193 @@ at::Tensor convolution_overrideable(
+      !params.is_dilated()) {
+    // fast path for grouped conv3d
+    return ConvBackend::Slow3d;
+  } else if (input.device().is_cpu() || input.is_cuda()) {
+    // backends without support for groups
+    if (params.transposed) {
+      if (input.ndimension() == 4) {
+        return ConvBackend::SlowTranspose2d;
+      } else if (input.ndimension() == 5) {
+        return ConvBackend::SlowTranspose3d;
+      } else {
+        // unsupported
+      }
+    } else {  /* Not transposed */
+      if (input.ndimension() == 4) {
+        if (params.is_dilated()) {
+          return ConvBackend::SlowDilated2d;
+        } else {  /* dim == 4, non-dilated */
+          if (params.use_nnpack(input, weight)) {
+            return ConvBackend::NnpackSpatial;
+          } else {
+            /* CPU implementation has specialized MM kernels
+               for non-dilated case here */
+            return ConvBackend::Slow2d;
+          }
+        }
+      } else if (input.ndimension() == 5 && (input.is_cuda() || params.is_dilated())) {
+        return ConvBackend::SlowDilated3d;
+      } else if (input.ndimension() == 5) { /* dim == 5, CPU, non-dilated */
+        /* CPU implementation has specialized MM kernels
+           for non-dilated case here */
+        return ConvBackend::Slow3d;
+      } else {
```
Where does this code come from? It doesn't seem to be in `_convolution`

zou3519 replied:
NVM found it in _convolution_nogroup

zou3519 commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -851,49 +1005,44 @@ at::Tensor _convolution(
+  // Select appropriate backend to use.
+  ConvBackend backend = select_conv_backend(input, weight, bias, params);
+
+  // Do some additional backend-specific checks.
+  if (backend == ConvBackend::Cudnn || backend == ConvBackend::CudnnTranspose ||
+      backend == ConvBackend::Miopen || backend == ConvBackend::MiopenTranspose) {
+    TORCH_CHECK(input.options().type_equal(weight.options()),
+        "Input type (", input.toString(), ") and weight type (", weight.toString(),
+        ") should be the same");
+    TORCH_CHECK(!bias.defined() || (input.options().type_equal(bias.options())),
+        "Input type (", input.toString(), ") and bias type (", bias.toString(),
+        ") should be the same");
+  }
```
nit: it might be nicer if the check were a helper function and then we could put the check in the switch statement below instead of special casing it up here

jbschlosser replied:
Sounds good, it might also be useful to apply this check more broadly, but I'm punting on that here since it's technically more restrictive.

zou3519 commented on test/test_nn.py:
```diff
@@ -13261,6 +13263,156 @@ def test_conv3d_valid_padding_backward(self, device):
+                decorators=[onlyCPU, skipCPUIfNoMkldnn], name='mkldnn_empty_batch_channel3d'),
+        # Note: Tests for mobile backends are not currently supported. This comprises
+        # NnpackSpatial, Winograd3x3Depthwise, and Xnnpack2d backends. Testing these
+        # requires the ability to gate tests by whether PyTorch is built with USE_MOBILE=1.
```
There appears to be `torch.backends.xnnpack.enabled`, not sure if that helps...

jbschlosser replied:
True, looks like I can get a test in for that backend by disabling cuDNN / mkldnn and checking `torch.backends.xnnpack.enabled`. My plan was to punt on mobile testing for this PR but I'll open an issue suggesting this

zou3519 reviewed (approved):
seems reasonable to me

albanD reviewed (approved):
SGTM only richard's nit

jbschlosser commented:
> > Some backends aren't tested yet:
> > miopen (for AMD ROCm)
> > nnpack (for mobile)
> > xnnpack (for mobile)
> > winograd 3x3 (for mobile)
> 
> I just started reading through this, but what is the plan for those? Is it fine to ignore them in this PR or do they need to be handled here?

Current plan is to ignore them for this PR but add a note that they should be handled at some point. I was able to get miopen tested, so that just leaves the mobile backends. I can open a mobile issue indicating that they're not tested and leave it to those folks to deal with it.

jbschlosser committed:
Factor backend routing logic out of convolution forward (#67790)

Summary:
This PR introduces a new function `_select_conv_backend` that returns a `ConvBackend` enum representing the selected backend for a given set of convolution inputs and params.

The function and enum are exposed to python for testing purposes through `torch/csrc/Module.cpp` (please let me know if there's a better place to do this).

A new set of tests validates that the correct backend is selected for several sets of inputs + params. Some backends aren't tested yet:
* nnpack (for mobile)
* xnnpack (for mobile)
* winograd 3x3 (for mobile)

Some flowcharts for reference:
![conv_routing_graph md](https://user-images.githubusercontent.com/75754324/140828957-1135b400-38c0-4c9f-87ef-4f33ceebeeae.png)
![conv_nogroup_routing_graph md](https://user-images.githubusercontent.com/75754324/140828977-ed223a4e-aa86-49f1-9925-c0f6b9ab36af.png)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/67790

Reviewed By: zou3519

Differential Revision: D32280878

Pulled By: jbschlosser

fbshipit-source-id: 4b9946f09411993b1c1d2f6bca8b485be823aebe
